### DISCUSSION TITLE
Start Keycloak Container in HTTPS

### DISCUSSION THREAD

**Discussion by damienbod** [microsoft/aspire#5497] [General]
I want to run the Keycloak container using HTTPS, not HTTP

Any ideas how to set this up?

```csharp

var keycloak = builder.AddKeycloak("keycloak", 8081)
    .WithDataVolume();
```

**Comment by davidfowl** [comment]
With lots of pain, generating a certificate, and wiring it up appropriately. @DamianEdwards has a sample https://github.com/dotnet/aspire-samples/blob/b741f5e78a86539bc9ab12cd7f4a5afea7aa54c4/samples/Keycloak/Keycloak.AppHost/KeycloakExtensions.cs#L18

**Reply by DamianEdwards** [reply]
That sample just reuses the ASP.NET Core dev cert rather than generating another, as it's likely to already be trusted. It exports it to a file using `dotnet  dev-certs` in a deterministic temp location, bind mounts it into the container, and uses the Keycloak container's existing support for supplying an HTTPS cert file path via environment variables to configure it.

**Reply by damienbod** [reply]
@DamianEdwards Works great thanks

I had to add an extra parameter to fix the port in the WithHttpsEndpoint method. 

I can make a PR in the samples if you want

Greetings Damien

**Reply by AkselKP** [reply]
Does this breaking change break this solution? 
https://learn.microsoft.com/en-us/dotnet/core/compatibility/aspnet-core/9.0/certificate-export

**Reply by DamianEdwards** [reply]
@AkselKP  The sample was updated to handle that https://github.com/dotnet/aspire-samples/blob/damianedwards/keycloak-sample/samples/Keycloak/Keycloak.AppHost/HostingExtensions.cs#L89

Note that the code in that branch has been refined and improved upon in other places including the [CommunityToolkit here](https://github.com/CommunityToolkit/Aspire/blob/main/src/Shared/DevCertHostingExtensions.cs). We're working on a first-class feature for this scenario (using the dev cert for non-project resources) which we hope to be included in the next version of Aspire.

**Comment by damienbod** [comment]
I have one HTTPS problem left. The Keycloak server (inside the container) needs to call the web applications in the host env (backchannel logout). The host.docker.internal is used for this. Do you now if it is possible to configure this to work with HTTPS in .NET Aspire? This will work in production, but dev would be cool as well :)

I am thinking of using ngrok, make a public URL and use this for the backchannel post request, but this is not really practical. 

Demo project: https://github.com/damienbod/keycloak-backchannel

Greetings Damien

**Comment by damienbod** [comment]
@rufer7 Provided this solution which works:

```csharp
var keycloak = builder.AddKeycloakContainer("keycloak",
            userName: userName, password: password, port: 8080)
    .WithArgs("--features=preview")
    .WithArgs("--spi-connections-http-client-default-disable-trust-manager=true")
    .WithDataVolume()
    .RunWithHttpsDevCertificate(port: 8081);
```

**Reply by rufer7** [reply]
Just make sure to **only** use this setting in local dev env!

For more details about `disable-trust-manager` setting see [here](https://www.keycloak.org/server/outgoinghttp#_client_configuration_command)

**Comment by LethargicDeveloper** [comment]
I've setup the Keycloak container to use HTTPS in Aspire using the following extension method

```
public static IResourceBuilder<KeycloakResource> WithHttpsEndpointWithCertificate(this IResourceBuilder<KeycloakResource> builder, int port)
{
    builder
        .WithHttpsEndpoint(env: "KC_HTTPS_PORT", port: port, targetPort: 8443)
        .WithEnvironment("KC_HTTPS_CERTIFICATE_FILE", "/certs/cert.pem")
        .WithEnvironment("KC_HTTPS_CERTIFICATE_KEY_FILE", "/certs/key.pem")
        .WithEnvironment("KC_HOSTNAME", "localhost")
        .WithEnvironment("KC_HTTP_ENABLED", "false")
        .WithEnvironment("QUARKUS_HTTP_HTTP2", "false")
        .WithBindMount("../../.cert", "/certs");

    return builder;
}
```

An HTTPS endpoint shows up in the aspire dashboard and that works. In my api project I've set up keycloak like this:

```
services.AddKeycloakWebApiAuthentication(
    builder.Configuration,
    auth =>
    {
        auth.Audience = "hypercube";
    });
    
services.AddOpenApiDocument(options =>
{
    options.Title = "Hypercube API";
    options.Version = "v1";
    options.DocumentName = "v1";

    var keycloakOptions = builder.Configuration.GetKeycloakOptions<KeycloakAuthenticationOptions>()!;

    options.AddSecurity("oauth2", new NSwag.OpenApiSecurityScheme
    {
        Type = NSwag.OpenApiSecuritySchemeType.OAuth2,
        Description = "OAuth2 Implicit Grant",
        OpenIdConnectUrl = $"{config["Keycloak:AuthServerUrl"]}/realms/hypercube/.well-known/openid-configuration",
        Flows = new NSwag.OpenApiOAuthFlows
        {
            Implicit = new NSwag.OpenApiOAuthFlow
            {
                AuthorizationUrl = keycloakOptions.KeycloakTokenEndpoint.Replace("/token", "/auth"),
                Scopes = new Dictionary<string, string>
                {
                    { "openid", "OpenID Connect scope" },
                    { "profile", "Profile scope" },
                    { "email", "Email scope" },
                }
            }
        }
    });

    options.OperationProcessors.Add(new AspNetCoreOperationSecurityScopeProcessor("oauth2"));
});
```

The problem is the Keycloak:AuthServerUrl and the TokenEndpointUrl still return the HTTP endpoint NOT the HTTPS endpoint. How do I get Keycloak to default to the configured HTTPS endpoint instead of HTTP? Or what is the appropriate way to configure that on the api side of things?

**Reply by DamianEdwards** [reply]
Where is `Keycloak:AuthServerUrl` coming from? Can you show the code from the app host that sets that?

**Comment by LethargicDeveloper** [comment]
I feel silly. I wasn't even using Microsoft's keycloak aspire package (Aspire.Hosting.Keycloak). I was using a different one (Keycloak.AuthServices.Aspire.Hosting). It looks like the AuthServerUrl is hardcoded to http in the library I'm using. Not sure you can switch it...

When I try switching to Aspire.Hosting.Keycloak I have other issues, probably related to the aspire workload. I'll just wait until it gets officially released before I switch over to it.

**Reply by DamianEdwards** [reply]
If it helps, there's an in-progress Keycloak you can reference at https://github.com/dotnet/aspire-samples/tree/damianedwards/keycloak-sample/samples/Keycloak

**Comment by gregpakes** [comment]
Is this still the recommended advice?  This is a little problematic on a Mac.

**Reply by DamianEdwards** [reply]
We're working on a new feature for Aspire 13.1 that provided first-class support for injecting the developer certificate into containers so that they can use them to host HTTPS endpoints. That would be the recommendation once it releases.

**Reply by davidfowl** [reply]
This now enabled by default in 13.1 with the dev cert.

**Reply by gregpakes** [reply]
Does it solve the issue on a mac with the keychain Allow prompt (that doesn't work)?

Namely when dotnet tries to export the cert, you get this prompt:

<img width="371" height="148" alt="image" src="https://github.com/user-attachments/assets/7433b08e-ef96-4922-acd5-ccd50a7bdac7" />

**Reply by davidfowl** [reply]
It needs to prompt at least once. It should work though (at leasst it works on my mac 😄 )

**Reply by davidfowl** [reply]
cc @danegsta

**Reply by gregpakes** [reply]
Mine prompts multiple times per app starting.  It doesn't matter if I click Always Allow.

The only way I can get rid of it was to do something in keychain, but I honestly can't remember what I did.  I know I had to do it again when i switched to .net 10.

So what you have in 13.1 is essentially the same process, just refined and "official"?

**Reply by davidfowl** [reply]
In a way that doesn't prompt multiple times per application.

**Reply by gregpakes** [reply]
Does it remember across application restarts?  Or will it prompt at least once everytime I run the application?

**Reply by davidfowl** [reply]
Yes its remembered across application restarts.

**Reply by danegsta** [reply]
Yeah, we’re caching the exported dev cert so that we aren’t having to trigger a keychain prompt every time we run. Worst case scenario is two prompts, as we have to prompt for each of the two certificate export formats (PFX and PEM) and if you have resources that need both you can end up being promoted twice.

**Reply by gregpakes** [reply]
Thanks - I already do a poor mans version of this caching...  I'll wait for 13.1 and feedback.  thanks.

**Reply by davidfowl** [reply]
You can try it now and give feedback before it ships

**Reply by gregpakes** [reply]
Apologies if I'm being slow - where is the 13.1 preview? I've checked nuget pre-release can can't find any preview packages?

**Reply by davidfowl** [reply]
https://github.com/dotnet/aspire/blob/main/docs/using-latest-daily.md

**Reply by danegsta** [reply]
This is now available in the stable Aspire 13.1 release.

**Reply by gregpakes** [reply]
@danegsta 

I can't even get the dashboard to load in 13.1.

In the browsere console I get:

```
Error: Grpc.Core.RpcException: Status(StatusCode="Internal", Detail="Error starting gRPC call. HttpRequestException: The SSL connection could not be established, see inner exception. AuthenticationException: The remote certificate is invalid because of errors in the certificate chain: UntrustedRoot", DebugException="System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.")
 ---> System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid because of errors in the certificate chain: UntrustedRoot
```

Here is my dev-cert info:

```
gregpakes@Mac ~ % dotnet dev-certs https --check -v --trust
[1] Listing certificates from CurrentUser\My
[2] Found certificates: 1 certificate
    1) 1FDF7793584C0C0A441A94E66449C6F66BB27BFD - CN=localhost - Valid from 2025-11-17 10:10:26Z to 2026-11-17 10:10:26Z - IsHttpsDevelopmentCertificate: true - IsExportable: true
[3] Checking certificates validity
[4] Valid certificates: 1 certificate
    1) 1FDF7793584C0C0A441A94E66449C6F66BB27BFD - CN=localhost - Valid from 2025-11-17 10:10:26Z to 2026-11-17 10:10:26Z - IsHttpsDevelopmentCertificate: true - IsExportable: true
[5] Invalid certificates: no certificates
[6] Finished listing certificates.
[23] Saving certificate '1FDF7793584C0C0A441A94E66449C6F66BB27BFD - CN=localhost - Valid from 2025-11-17 10:10:26Z to 2026-11-17 10:10:26Z - IsHttpsDevelopmentCertificate: true - IsExportable: true' to /var/folders/fw/t8j0txw56sq1j9597g9brsh40000gn/T/tmpsPPI2J.tmp without private key.
[25] Creating directory /var/folders/fw/t8j0txw56sq1j9597g9brsh40000gn/T.
[27] Writing the certificate to: /var/folders/fw/t8j0txw56sq1j9597g9brsh40000gn/T/tmpsPPI2J.tmp.
A trusted certificate was found: 1FDF7793584C0C0A441A94E66449C6F66BB27BFD - CN=localhost - Valid from 2025-11-17 10:10:26Z to 2026-11-17 10:10:26Z - IsHttpsDevelopmentCertificate: true - IsExportable: true
```

**Reply by DamianEdwards** [reply]
Which browser are you using? You can try cleaning and recreating your certs:
```shell
> dotnet dev-certs https --clear
> dotnet dev-certs https --trust
```

**Reply by gregpakes** [reply]
I did that - Issue persists.

Using Edge on Mac 26.1.

**Reply by gregpakes** [reply]
Reboot fixed....  🤦‍♂️

**Comment by gregpakes** [comment]
I have another question about this in 13.1.

I can use the new 13.1 bits to get all the services being served using the developer certificate and thats great. 

My issue now is that in Keycloak specifically, this doesn't seem to put the cert in the JVM's truststore.  So when keycloak tries to verify the certificate of my IDP (DEX), I get:

```
2026-01-16 14:42:14,378 ERROR [org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider] (executor-thread-3) Failed to make identity provider oauth callback: javax.net.ssl.SSLHandshakeException: (certificate_unknown) PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:130)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:383)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:326)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:321)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1287)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1164)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1107)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:393)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:477)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:448)
	at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:206)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:172)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1506)
	(truncated)
```

Is this expected?

I tried the `.WithDeveloperCertificateTrust(true);` extension method, but that didn't work either.

Edit:

I managed to fix this with:

``` c#
.WithHttpsCertificateConfiguration(ctx =>
{
    ctx.EnvironmentVariables["KC_TRUSTSTORE_PATHS"] = ctx.CertificatePath;
    return Task.CompletedTask;
});
```
